### DISCUSSION TITLE
Feature for writeContract function

### DISCUSSION THREAD

**Discussion by glitch-txs** [wevm/viem#757] [Idea / Feature Request] [answered]
Hi, as we know `writeContract` fn takes a chain as optional param and it will through an error if there's a mismatch between the wallet's currently chain and the configured one.
not sure if this is wise but talking on a frontend perspective it would be really cool if, instead of just throwing an error it would trigger `eth_switchEthereumChain` and if the chain is not added `eth_addEthereumChain` and then retrigger the tx function or throw an error if the requests were rejected.

This is something I would definitely try to implement on any transaction on a dapp, in the case of WalletConnect's ethereum-provider, the switchEthereumChain method would automatically change the default chain so it would be very useful there as well.

Not sure if this would make sense on backend or other environments though.

**Comment by jxom** [comment, marked as answer]
I am a bit hesitent to do this for a couple of reasons:

1. This breaks viem's [Rule of Transparency](http://www.catb.org/~esr/writings/taoup/html/ch01s06.html#:~:text=else%20will%20do.-,Rule%20of%20Transparency%3A%20Design%20for%20visibility%20to%20make%20inspection%20and%20debugging%C2%A0easier.,-Rule%20of%20Robustness): we want to build transparent APIs with no internal side-effects – these get incredibly hard for consumers to debug & manage error cases if they don't know what's going on internally.
2. We don't know if the Wallet supports `wallet_switchEthereumChain` or `wallet_addEthereumChain`, and the absense of those methods is a very common scenario.
3. It ends up affecting performance: if we are performing asynchronous RPC procedures from within the Action, we have no idea how long the Wallet will take to respond, and consequently `eth_sendTransaction` may be invoked several hundreds of milliseconds after the user invokes `writeContract`, which is not ideal. 

I think the most ideal scenario would that we would be able to pass a `chainId` in RPC payloads, but that is obviously not possible. :P

I think it's best to leave viem unopinionated and folks (e.g. wagmi) can build the opinionated layers on top of this.

**Reply by glitch-txs** [reply]
Makes sense 👌
